python/tests/ibd.py

```diff
@@ -0,0 +1,421 @@
+
+    def get_sample_id_map(self):
+        """
+        Returns id_map, a vector of length ts.num_nodes. For a node with id i,
+        id_map[i] is the position of the node in self.samples. If i is not a
+        user-specified sample, id_map[i] is -1.
+        Note: it assumes nodes in the TS are numbered 0 to ts.num_nodes - 1
+        """
+        id_map = [-1] * self.ts.num_nodes
+        for i, samp in enumerate(self.samples):
+            id_map[samp] = i
+
+        return id_map
+
+    def get_sample_pairs(self):
+        """
+        Returns a list of all pairs of samples. Replaces itertools.combinations.
+        Note: they must be sorted
+        """
+        sample_pairs = []
+        for ind, i in enumerate(self.samples):
+            for j in self.samples[(ind + 1) :]:
+                sample_pairs.append((i, j))
+
+        return sample_pairs
+
+    def find_sample_pair_index(self, sample0, sample1):
+        """
+        Note: this method isn't strictly necessary for the Python implementation
+        but is needed for the C implemention, where the output ibd_segments is a
+        struct array.
+        This calculates the position of the object corresponding to the inputted
+        sample pair in the struct array.
+        """
+
+        # Ensure samples are in order.
+        if sample0 == sample1:
+            raise ValueError("Samples in pair must have different node IDs.")
+        elif sample0 > sample1:
+            sample0, sample1 = sample1, sample0
+
+        i0 = self.sample_id_map[sample0]
+        i1 = self.sample_id_map[sample1]
+
+        # Calculate the position of the sample pair in the vector.
+        index = (
+            (self.num_samples) * (self.num_samples - 1) / 2
+            - (self.num_samples - i0) * (self.num_samples - i0 - 1) / 2
+            + i1
+            - i0
+            - 1
+        )
+
+        return int(index)
+
+    def find_ibd_segments(self, min_length=0, max_time=None):
+        """
+        The wrapper for the procedure that calculates IBD segments.
+        """
+
+        # Set up an iterator over the edges in the tree sequence.
+        mygen = iter(self.ts.edges())
+        e = next(mygen)
+        nodes_to_process = []
+
+        # Iterate over the edges.
+        while e is not None:
+            # Process all edges with the same parent node.
+            self.current_parent = e.parent
+            self.current_time = self.tables.nodes.time[self.current_parent]
+            if self.max_time is not None and self.current_time > self.max_time:
+                # Stop looking for IBD segments once the
+                # processed nodes are older than the max time.
+                break
+
+            seg = Segment(e.left, e.right, e.child)
+
+            # Create a SegmentList() holding all segments that descend from seg.
+            list_to_add = SegmentList()
+            u = seg.node
+            if u in self.samples:
```

Comment:
I'm confused here: why isn't `u` always `None`, since `node` wasn't given in the call to `Segment( )` above?

Reply:
did you mean to have `node=e.parent`?

Reply:
oh wait I'm being dumb sorry